### ISSUE TITLE
Provide an easy way to select the version of Flutter to use

### ISSUE THREAD

**Issue by rock3r** [flutter/flutter#14230]
This ties in with #14229 — it would be great if there was a way to select the version of Flutter to use from the `flutter` tool. Currently to my understanding the manual process to do so is:

```bash
$ cd flutter
$ git checkout [branch, tag or commit hash]
$ cd ../my-app/
$ flutter packages get
```

(assuming both the `flutter` folder and the `my-app` folder live in the same directory)

What would be nice would be able to use a command like `flutter switch-to [version]` (e.g., `flutter switch-to 0.0.20`). Ideally running `flutter run` or `flutter build` could do this automatically, but that poses the question of how it would work for people working on multiple projects using different versions of Flutter at the same time.

**Comment by rock3r** [comment]
CC @devoncarew

**Comment by Hixie** [comment]
We have `flutter channel <branch>`, where branch is master, dev, or beta. We don't have a way to explicitly pick an old version. The problem with supporting that is that you're likely to switch to a version that doesn't support that command, and then you can't get back. So instead we encourage people to use "git checkout" (as above) if they want a very specific version.

**Comment by rock3r** [comment]
I see your point; on the other hand, the sooner this is implemented the lesser that risk is. And I imagine once Flutter goes to a more "stable" setting such as betas or RCs or 1.x, it would become increasingly important to be able to easily enforce (and switch to) the version of Flutter a project was built with. I'm thinking more long-term than shorter term. You could easily put a lower bound on the version the command can downgrade too, as a safety net.

**Comment by Hixie** [comment]
Yeah, having a lower bound would make sense for this.

**Comment by RafaO** [comment]
I think this is a good idea, or at least specify the version of flutter SDK that should be used. I see also a nice benefit here for shared projects where we need to make sure everyone uses the same flutter version.

**Comment by rrousselGit** [comment]
This can probably be closed now that we can do `flutter version v1.2.3`

**Comment by DanTup** [comment]
Yep, this was implemented in https://github.com/flutter/flutter/pull/26840. Thanks!

**Comment by tuanbs** [comment]
> This can probably be closed now that we can do `flutter version v1.2.3`

I got the error `Unable to checkout version branch for version <version>` using cmd `flutter version <version>`.

**Comment by jibbers42** [comment]
> I got the error `Unable to checkout version branch for version <version>` using cmd `flutter version <version>`.

Me too. The flutter tool doesn't seem to download git tags properly. I'm not sure what its trigger is to get them, but I had to go into the flutter directory and do `git pull --tags origin stable`. I guess you'd replace `stable` with whichever channel you're on.

**Comment by migalv** [comment]
I seems like I'm up to date, but I still cannot change versions.

```
PS D:\migal\workspace\flutter> flutter version v1.9.1+hotfix.6
Unable to checkout version branch for version 1.9.1+hotfix.6.
PS D:\migal\workspace\flutter> git pull --tags origin stable
From https://github.com/flutter/flutter
 * branch                stable     -> FETCH_HEAD
Already up to date.
PS D:\migal\workspace\flutter> git pull --tags origin       
Already up to date.
PS D:\migal\workspace\flutter> git fetch
PS D:\migal\workspace\flutter> flutter version v1.9.1+hotfix.6
Unable to checkout version branch for version 1.9.1+hotfix.6.
```

any ideas??

**Comment by ngothanhtai** [comment]
> > This can probably be closed now that we can do `flutter version v1.2.3`
> 
> I got the error `Unable to checkout version branch for version <version>` using cmd `flutter version <version>`.

Make sure you switch to the right channel to get the versions.
For example: there are versions that are available on `master` instead of `stable` branch. Using `flutter channel master` to switch to master branch.

**Comment by samandmoore** [comment]
Hi there, I see this is closed, but I have some questions about the intended behavior.

Is it supposed to be possible to run `flutter version vX.X.X` from inside the flutter install directory? I'm currently using flutter beta channel (v1.14.6) and when I run `flutter version v1.13.6` from the flutter installation directory, I get the following error:

```
sam 〉~/flutter 〉beta
$ flutter version
/Users/sam/flutter: no pubspec.yaml found
```

If I try to run the command from a project directory, I get a different error:

```
sam 〉~/src/my_app
$ flutter version v1.13.6 -h
The current Flutter SDK version is 1.14.6.

Because my_app requires Flutter SDK version 1.13.6, version solving failed.
Running "flutter pub get" in my_app...
pub get failed (1; Because my_app requires Flutter SDK version 1.13.6, version solving failed.)
```

Thanks!

**Comment by stuz5000** [comment]
I'm confused. If I have:
```
$ flutter doctor -v
[✓] Flutter (Channel master, v1.15.4-pre.66, on Linux 4.15.0-74-generic, locale en_US.UTF-8)
    • Flutter version 1.15.4-pre.66 at /home/sir/development/flutter
    • Framework revision b34046903b (4 hours ago), 2020-02-13 18:34:08 -0800
    • Engine revision 0e8ed2715a
    • Dart version 2.8.0 (build 2.8.0-dev.8.0 5244d99a5d)
```
How can I get to the same version? If I do
```
flutter channel master
git checkout v1.15.4-pre.66
```
or 
```
git checkout v1.15.4-pre.66
```
I get:

```
error: pathspec 'v1.15.4-pre.66' did not match any file(s) known to git.
```

If I do:
```
$ flutter version v1.15.4-pre.66
Running "flutter pub get" in flutterui...                           0.5s
╔══════════════════════════════════════════════════════════════════════════════╗
║ Warning: "flutter version" will leave the SDK in a detached HEAD state.      ║
║ If you are using the command to return to a previously installed SDK version ║
║ consider using the "flutter downgrade" command instead.                      ║
╚══════════════════════════════════════════════════════════════════════════════╝


Are you sure you want to proceed? [y|n]: y
There is no version: 1.15.4-pre.66
Unable to checkout version branch for version 1.15.4-pre.66.
```
... same with  'v1.15.4-pre.66'

**Comment by matthew-carroll** [comment]
@Hixie @jonahwilliams @dnfield I'm seeing a number of issues posted to this closed ticket. I also just had similar confusion/issue with trying to use `flutter version v1.13.6` - it didn't do anything for me. I think either we could use some education here, or maybe there is a problem that needs to be fixed in the tooling. Can one of you comment?

In my case I ended up going to my flutter distro and I manually checked out the desired tag.

**Comment by dnfield** [comment]
I believe @zanderso and @jonahwilliams have some plans to rework and/or deprecate the version command into something more useable.

**Comment by matthew-carroll** [comment]
If it truly doesn't work in its current form, would it make sense to deprecate now, point out that it's buggy, and remind people of the workaround via git? 

I'm not sure it's doing anyone any good in it's current state...

**Comment by jonahwilliams** [comment]
Yes, this command as implemented doesn't really work. The best way to get on a particular version of Flutter is to either use the right git commands to select it, or download it from the website.

Re-opening the issue seems like the right thing to do here.

**Comment by jmagman** [comment]
Might be better after https://github.com/flutter/flutter/pull/52141/

**Comment by dnfield** [comment]
I didn't make #52141 apply to the `version` command, but could.  If we're keeping it around then it would make sense.

**Comment by zanderso** [comment]
@jonahwilliams @dnfield If possible can we close this issue in favor of (a) more specific issue(s) that describe what work we could do for this?

**Comment by Pradeep-ShareChat** [comment]
- Checkout flutter stable version from git repo
- If you are still getting error means some of your pubspec.yaml files are not updated w.r.t. latest flutter
- Find the version of flutter where it can be stable
- In my case i had to checkout `git checkout v1.12.13+hotfix.5`

**Comment by deliciafernandes** [comment]
**To switch a flutter chanel** 
- `flutter channel channel_name` eg. `flutter channel stable`
- `flutter upgrade`

**To change a flutter version**
-  `flutter downgrade flutter_version_name` eg. `flutter downgrade v1.17.0`

**Comment by vanlooverenkoen** [comment]
I have written a fastlane private lane that can be used for making sure every build uses the correct flutter version

**Comment by normancarcamo** [comment]
> **To switch a flutter chanel**
> 
> * `flutter channel channel_name` eg. `flutter channel stable`
> * `flutter upgrade`
> 
> **To change a flutter version**
> 
> * `flutter downgrade flutter_version_name` eg. `flutter downgrade v1.17.0`

Didn't work for me, it says: `There is no previously recorded version for channel "master".` when I perform: `flutter downgrade 1.20.0`

`flutter version`
```sh
Flutter 1.22.0-10.0.pre.39 • channel master • https://github.com/flutter/flutter.git
Framework • revision a8281e31af (24 hours ago) • 2020-09-01 10:57:59 -0700
Engine • revision 165abef0a2
Tools • Dart 2.10.0 (build 2.10.0-77.0.dev)

```

**Comment by AresVampire** [comment]
the same for me

**Comment by normancarcamo** [comment]
I opened a new issue for my case, if anybody runs into the same, then [follow this thread](https://github.com/flutter/flutter/issues/65175)

**Comment by vashisth00** [comment]
Guys not flutter version doesn’t work only flutter downgrade does

**Comment by Jonas-Sander** [comment]
Man how is this still not implemented. I need to change versions often to build mobile, desktop, web etc. and this is just a pain in the ***. I don't want to search for the right commit, I just want to change the version.

**Comment by jmagman** [comment]
[`flutter version` was deprecated](https://github.com/flutter/flutter/issues/53364) as of https://github.com/flutter/flutter/pull/67478.

You can run `flutter upgrade` and `flutter downgrade` on known channels (`dev`, `beta`, `stable`).  To select a particular version, from your Flutter SDK repo run `git checkout <version>` like `git checkout 1.22.6`.  If you are on the `master` channel you should check out the commit SHAs directly as versions ending in `.pre.#` are not actually git tags but are calculated as the number of commits since the version.

**Comment by Jonas-Sander** [comment]
Yeah I understand but as this issue is still open I was hoping for a replacement. 

> To select a particular version, from your Flutter SDK repo run git checkout <version> like git checkout 1.22.6.

I didn't knew this, I was always looking for a specific SHA, thanks!
Still I would wish for this type of command in the flutter CLI - especially for me I would wish for something which is smart enough to get the closest version of whatever I type in. E.g.: given the input `1.23.0` for the beta channel I would like to automatically get the flutter version `1.23.0-18.1.pre`. Even better would be an interactive CLI where I could just choose from a list of versions and automatically switch. I tried something like `fvm` but had problems with it.

**Comment by dnfield** [comment]
The major problem with a CLI command for this is that it's far more complicated to build correctly than it would first seem.

The following are all problems that have to be addressed:

- What do we do if there are unstaged changes?
- Do we grab all tags from upstream every time, which can be slow for people with slow internet connections?
  - If not, what do we do if you request to switch to an invalid version even though there's a valid version upstream?
- What happens if you switch to a version that doesn't support the command anymore?
- What happens if you switch to a version that has slightly different defaults/requirements for the command?
- Who will maintain this feature as other bugs/feature requests related to it arise?

IMHO, it is better to try to resolve these problems outside of the Flutter CLI itself, e.g. with something like the Flutter Version Manager (FVM), or flutter_wrapper. I suspect these solutions are imperfect, but it's easier to evolve them separately from the tool and its requirements and avoid some of the problems I mentioned above completely.

I think we should close this bug as wont-fix, but maybe @jonahwilliams or @zanderso have other plans for it.

**Comment by dnfield** [comment]
(of course, another option to solve it is the one @jmagman mentioned, just directly using git commands yourself in the repo)

**Comment by Wissperwind** [comment]
Hey,
I think the important function to upgrade to a specific version is still missing!
We can downgrade to a specific version, but what if I would like to update to a version that is not the newest?

At the moment I make our build process. If I release a new version, the build server needs to upgrade flutter to a new version that is not necessarily the newest version but the version used in development at the moment.

**Comment by Wissperwind** [comment]
To my understanding, at the moment there is no option to make the build server install for example:
2.4.0-1.0.pre.148
What could be the current version used in development.
To my furthe runderstanding, the only option I have is to get the git commit from a dev computer:
```
git rev-parse HEAD
633af3da0ec6eb95ad8959f80d24a5d3fb3aed50
```
And that make the build server install this by using git checkput and flutter docor. The git commands must be executed in the sdk folder.
Ok - at least there is an option.

**Comment by illia-romanenko** [comment]
I'm not sure why `flutter channel` can be implemented correctly (as it probably relies on git commands too), but `flutter version` can't be, won't be? 

I think the idea here should be is how easy it's to switch to a specific flutter version for devs if they need to - and I believe that `flutter version` is much simpler than looking for flutter installation and doing git commands there. Especially that you still need to use `flutter` command to download dependencies after the switch (which is annoying and confusing for newcomers too).

**Comment by dnfield** [comment]
There are at least two community based solutions to address this, and it's possible to do via the CLI in ways that don't bring up all the problems I mentioned in https://github.com/flutter/flutter/issues/14230#issuecomment-771173468

I'm going to close this as wontfix.

**Comment by stuz5000** [comment]
🤦‍♂️ It's not like getting up and running is a stable environment isn't
top of people’s wish lists.
I personally burned around a day over a year here - bugs in flutter make
searching for an actual stable release a requirement.

**Comment by christopherfujino** [comment]
I'm re-opening this; though I know of no immediate plans to implement this, I think it would be worth doing.

**Comment by lukepighetti** [comment]
I don't see any way to bypass the user confirmation. I was hoping `flutter downgrade 2.5.3 --force` would get me there

**Comment by christopherfujino** [comment]
> I don't see any way to bypass the user confirmation. I was hoping `flutter downgrade 2.5.3 --force` would get me there. Which means I cannot use this in a script AFAIK?

From looking at the source, it looks like you can provide `--no-prompt` to disable it.

**Comment by Wissperwind** [comment]
That looks cool. Maybe I can upgrade out build server to finally install a specific flutter version for the build.
But there I tried and I have one question:
```
C:\Users\myUser>flutter downgrade 2.5.3 --no-prompt
Downgrading Flutter to version 2.10.0-0.0.pre-256-g5e2542f609
```
If I request version 2.5.3: Why does he load Version 2.10?
Are there different nummerations?

**Comment by lukepighetti** [comment]
If that's true, then picking versions is really broken. There should be a way to do this in CI without having to specify a commit hash

**Comment by christopherfujino** [comment]
> If that's true, then picking versions is really broken. There should be a way to do this in CI without having to specify a commit hash

you can checkout a git tag

**Comment by Wissperwind** [comment]
At the moment I use git hashes. Is it tested that going forward and backward by using git hashes always works?
I run git chackout and afterwards flutter doctor. I remember but can't proof it that one time going forward id didn't work.
What makes me wonder: If I run flutter upgrade many things, more than ten, were installed. But if I run git checkout and flutter doctor only a few things were installed.

If I do git checkout and afterwards flutter doctor to install an arbitrary version, will it install all of the following things correctly to that version?
```
Downloading android-arm-profile/windows-x64 tools...             1.014ms
Downloading android-arm-release/windows-x64 tools...               705ms
Downloading android-arm64-profile/windows-x64 tools...             785ms
Downloading android-arm64-release/windows-x64 tools...             848ms
Downloading android-x64-profile/windows-x64 tools...               848ms
Downloading android-x64-release/windows-x64 tools...               746ms
Downloading android-x86 tools...                                 2.046ms
Downloading android-x64 tools...                                 1.945ms
Downloading android-arm tools...                                 2.795ms
Downloading android-arm-profile tools...                         1.135ms
Downloading android-arm-release tools...                           752ms
Downloading android-arm64 tools...                               2.161ms
Downloading android-arm64-profile tools...                       1.157ms
Downloading android-arm64-release tools...                         957ms
Downloading android-x64-profile tools...                         1.205ms
Downloading android-x64-release tools...                           857ms
Downloading android-x86-jit-release tools...                     1.269ms
Downloading Web SDK...                                              5,1s
Downloading CanvasKit...                                         1.547ms
Downloading package sky_engine...                                  342ms
Downloading flutter_patched_sdk tools...                           685ms
Downloading flutter_patched_sdk_product tools...                   666ms
Downloading windows-x64 tools...                                    4,7s
Downloading windows-x64/windows-x64-flutter tools...                9,0s
Downloading windows-x64/flutter-cpp-client-wrapper tools...        510ms
Downloading windows-x64-profile/windows-x64-flutter tools...         7,9s
Downloading windows-x64-release/windows-x64-flutter tools...         7,4s
Downloading windows-x64/font-subset tools...                       563ms
```

These are the things installed when running flutter upgrade.

**Comment by lukepighetti** [comment]
> you can checkout a git tag

And what if your installation is not the flutter git repo?

**Comment by zanderso** [comment]
> > you can checkout a git tag
> 
> And what if your installation is not the flutter git repo?

@lukepighetti Could you describe a bit more about what you have in mind?

**Comment by christopherfujino** [comment]
> At the moment I use git hashes. Is it tested that going forward and backward by using git hashes always works? I run git chackout and afterwards flutter doctor. I remember but can't proof it that one time going forward id didn't work. What makes me wonder: If I run flutter upgrade many things, more than ten, were installed. But if I run git checkout and flutter doctor only a few things were installed.
> 
> If I do git checkout and afterwards flutter doctor to install an arbitrary version, will it install all of the following things correctly to that version?
> 
> ```
> Downloading android-arm-profile/windows-x64 tools...             1.014ms
> Downloading android-arm-release/windows-x64 tools...               705ms
> Downloading android-arm64-profile/windows-x64 tools...             785ms
> Downloading android-arm64-release/windows-x64 tools...             848ms
> Downloading android-x64-profile/windows-x64 tools...               848ms
> Downloading android-x64-release/windows-x64 tools...               746ms
> Downloading android-x86 tools...                                 2.046ms
> Downloading android-x64 tools...                                 1.945ms
> Downloading android-arm tools...                                 2.795ms
> Downloading android-arm-profile tools...                         1.135ms
> Downloading android-arm-release tools...                           752ms
> Downloading android-arm64 tools...                               2.161ms
> Downloading android-arm64-profile tools...                       1.157ms
> Downloading android-arm64-release tools...                         957ms
> Downloading android-x64-profile tools...                         1.205ms
> Downloading android-x64-release tools...                           857ms
> Downloading android-x86-jit-release tools...                     1.269ms
> Downloading Web SDK...                                              5,1s
> Downloading CanvasKit...                                         1.547ms
> Downloading package sky_engine...                                  342ms
> Downloading flutter_patched_sdk tools...                           685ms
> Downloading flutter_patched_sdk_product tools...                   666ms
> Downloading windows-x64 tools...                                    4,7s
> Downloading windows-x64/windows-x64-flutter tools...                9,0s
> Downloading windows-x64/flutter-cpp-client-wrapper tools...        510ms
> Downloading windows-x64-profile/windows-x64-flutter tools...         7,9s
> Downloading windows-x64-release/windows-x64-flutter tools...         7,4s
> Downloading windows-x64/font-subset tools...                       563ms
> ```
> 
> These are the things installed when running flutter upgrade.

The reason you see more things downloaded when doing `flutter upgrade` is because under the hood, it does a `flutter precache` after checking out the new version: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/upgrade.dart#L206. However, this is just an optimization; if you subsequently call a command that requires a binary artifact that has not been downloaded, the tool will first download that, and then run the requested command.

**Comment by neoacevedo** [comment]
Having the version 2.8.1, when I run _flutter version_ I get:
```sh
#:~/flutter version
Could not find a command named "version".

Run 'flutter -h' (or 'flutter <command> -h') for available flutter commands and options.
```

**Comment by jmagman** [comment]
> [`flutter version` was deprecated](https://github.com/flutter/flutter/issues/53364) as of https://github.com/flutter/flutter/pull/67478.
> 
> You can run `flutter upgrade` and `flutter downgrade` on known channels (`dev`, `beta`, `stable`).  To select a particular version, from your Flutter SDK repo run `git checkout <version>` like `git checkout 1.22.6`.  If you are on the `master` channel you should check out the commit SHAs directly as versions ending in `.pre.#` are not actually git tags but are calculated as the number of commits since the version.

This issue is tracking a replacement for `flutter version` that doesn't require git commands.

**Comment by mhmzdev** [comment]
Anything on this yet?
The `flutter version <some-version` isn't working. Using flutter 3.x.x

**Comment by matanlurey** [comment]
We don't imagine adding a new way to do this outside of checking out tags.

If the release strategy changes in the future (i.e. does not use git) then of course this will change.
